### PR TITLE
PR: Create a separate window when undocking plugins

### LINKED ISSUE #3563: Docks confusion when pressing undocking button
Some feedback from a lecture @jitseniesen gave recently (October 2016)

> Somebody complained that the dock window interface is confusing; the undock button looks like it would maximize, and he got confused with the undocked window because it does act like a normal window does (this is on Linux)

### LINKED ISSUE #3790: Create new windows when undocking all plugins
There are two improvements I'd like to see for new Editor windows

1. Make their toolbar to be exactly the same as the one present in the main window (except for the working directory toolbar, which we don't need to show in new windows).
2. When the Editor plugin is undocked, we must close it and show a new window instead. For that we need to connect a new method to the topLevelChanged signal of its QDockWidget:

    http://doc.qt.io/qt-4.8/qdockwidget.html#topLevelChanged

@dalthviz, please work on this in two different PRs. Please start with point 1. which is easier, and then move to 2.

### PR DESCRIPTION
**Pull request by dalthviz**
Fixes #3790 
Fixes #3563

### CODE REVIEW

**Comment by ccordoba12** [comment]
Please update with the latest changes in the 3.x branch to fix the errors in CircleCI

File: spyder/app/mainwindow.py [hunk 1]
```diff
@@ -2209,6 +2210,15 @@ def __update_fullscreen_action(self):
         self.fullscreen_action.setIcon(icon)
 
     @Slot()
+    def undock_editor(self):
+        """Open a new window instance of the Editor instead of undoking it."""
```
Comment (ccordoba12) [inline spyder/app/mainwindow.py]: `undoking` is misspelled here. It's `undocking`

**Comment by ccordoba12** [comment]
There's one thing that worries me about this change. What happens when users press `Ctrl+Shift+E` to move to the Editor?

I think that makes the Editor pane to appear in the main window, right?

**Comment by dalthviz** [comment]
If you are with the focus in the main window, yes it makes appear the Editor pane. If you, in the other hand, have the focus in a new window it doesn't appear.

**Comment by ccordoba12** [comment]
> If you are with the focus in the main window, yes it makes appear the Editor pane

Then we need to fix this, and make `Ctrl+Shift+E` to move to the new window, if the pane is hidden and an Editor window is visible.

We also need to prevent to more than window for this to make sense :-)

**Comment by ccordoba12** [comment]
Sorry @dalthviz for my big delay on reviewing this one. I hope to do it after we release 3.1 :-)

**Comment by goanpeca** [comment]
@dalthviz please rebase or merge with the latest 3,x to fix conflicts

File: spyder/app/mainwindow.py [hunk 2]
```diff
@@ -2240,6 +2241,17 @@ def __update_fullscreen_action(self):
...
     @Slot()
+    def undock_editor(self):
+        """Open a new window instance of the Editor instead of undocking it."""
+        if self.editor.dockwidget.isFloating() and not self.editor.undocked:
+            self.editor.dockwidget.setVisible(False)
+            self.editor.create_new_window()
+            self.editor.toggle_view_action.setChecked(False)
+            self.editor.dockwidget.setFloating(False)
+        self.editor.undocked = False
+        self.editor.get_current_editorstack().new_window = False
```
Comment (ccordoba12) [inline spyder/app/mainwindow.py]: @dalthviz, does this method need to be here in the main window? As far as I can see, it doesn't uses any attribute/method of the window, so I think we can safely move it to the the Editor plugin file.

File: spyder/app/mainwindow.py [hunk 3]
```diff
@@ -799,6 +799,7 @@ def create_edit_action(text, tr_text, icon):
         from spyder.plugins.editor import Editor
         self.editor = Editor(self)
         self.editor.register_plugin()
+        self.editor.dockwidget.topLevelChanged.connect(self.undock_editor)
```
Comment (ccordoba12) [inline spyder/app/mainwindow.py]: This line can also be moved to the Editor plugin. We can make it the last one of its `register_plugin` method.

**Comment by ccordoba12** [comment]
@dalthviz, please merge with 3.x and finish this one for Spyder 3.2. I think it's almost ready ;-)

**Comment by ccordoba12** [comment]
@dalthviz, please create an animated gif so that @goanpeca and me could see how this is working.

**Comment by dalthviz** [comment]
Preview:

undock

**Comment by ccordoba12** [comment]
@goanpeca, what do you think? There is an action in the Editor options menu to undock it so users can move next to other plugins.

**Comment by ccordoba12** [comment]
Problem is this is not the same behavior for the rest of plugins.

**Comment by ccordoba12** [comment]
@dalthviz, this needs another merge with `3.x`.

**Comment by goanpeca** [comment]
@ccordoba12 yeah, this change would be a bit inconsistent with the rest and is better suited for 4.x

**Comment by ccordoba12** [comment]
Ok, you're right. @dalthviz, please rebase this PR in top of master and keep working on it, so that when users press the undock button they see a new window for *every* plugin, and not just the Editor.

This functionality could go to `plugins/base.py` in master.

**Comment by dalthviz** [comment]
@ccordoba12 @goanpeca When working in this PR I found that some plugins (atleast 8) doesn't have a toolbutton to add the undock action. What should we do?

**Comment by ccordoba12** [comment]
Our decision is that (for consistency) all plugins must have an Options menu, that should contain at least the undocking entry.

File: spyder/api/plugins.py [hunk 4]
```diff
@@ -311,3 +312,17 @@ def check_compatibility(self):
         message = ''
         valid = True
         return valid, message
+
+class PluginMainWindow(QMainWindow):
```
Comment (ccordoba12) [inline spyder/api/plugins.py]: Please move this to `plugins/base.py`

**Comment by ccordoba12** [comment]
@rlaverde, please review this one.

**Comment by dalthviz** [comment]
@ccordoba12 @rlaverde a preview:

undocking

Also, while doing the preview I realized that we should need to verify if the third party plugins need some ajustment to fit with the changes in here. As an example the `unit-test` plugin:

imagen

Should I create and resolve in each repository an issue related with this?

**Comment by ccordoba12** [comment]
> Should I create and resolve in each repository an issue related with this?

Yep, please open an issue in all our plugins that require a fix.

**Comment by ccordoba12** [comment]
By the way, great @dalthviz! This is looking really, really nice!

**Comment by rlaverde** [comment]
I just tested locally and It looks great, maybe the only think that I don't like is when undocking plugins are somehow small and square; maybe they should be rectangular and a little bigger, or use the dock dimensions.

**Comment by ccordoba12** [comment]
> the only think that I don't like is that when undocking plugins are somehow small and square

Could you post a screenshot to see what you mean?

**Comment by rlaverde** [comment]
>Could you post a screenshot to see what you mean?

If you see the video that @dalthviz posted, after undocking the plugin is necessary to resize the window because it is somehow small

File: spyder/widgets/dock.py [hunk 5]
```diff
@@ -153,10 +153,10 @@ def show_nontab_menu(self, event):
 
 class SpyderDockWidget(QDockWidget):
     """Subclass to override needed methods"""
-    plugin_closed = Signal()
+    plugin_closed = Signal(object)
 
     def __init__(self, title, parent):
-        super(SpyderDockWidget, self).__init__(title, parent)
+        super(SpyderDockWidget, self).__init__(title)
```
[the remark this answers is not in the document]
Comment (dalthviz) [reply spyder/widgets/dock.py]: I think this change is not necessary.  I will revert this :+1:

File: spyder/plugins/base.py [hunk 6]
```diff
@@ -169,3 +172,53 @@ def toggle_view(self, checked):
...
+    def create_mainwindow(self):
+        """
+        Create a QMainWindow instance containing this plugin.
+        """
+        raise NotImplementedError
+
+    def create_undock_action(self):
+        """Create the undock action for the plugin."""
+        self.undock_action = create_action(self,
+                                           _("Undock the plugin"),
```
Comment (ccordoba12) [inline spyder/plugins/base.py]: Let's change this to simply `Undock`, instead of `Undock the plugin`.

File: spyder/plugins/base.py [hunk 7]
```diff
@@ -169,3 +172,53 @@ def toggle_view(self, checked):
...
+                                           triggered=self.undock_plugin)
+
+    def undock_plugin(self):
+        """Undocks the plugin from the MainWindow."""
+        self.undocked = True
+        self.dockwidget.setFloating(True)
+        self.undock_action.setDisabled(True)
+
+
+class PluginMainWindow(QMainWindow):
```
Comment (ccordoba12) [inline spyder/plugins/base.py]: Please move this class to be before `BasePluginWidget`

File: spyder/api/plugins.py [hunk 8]
```diff
@@ -66,6 +68,10 @@ def __init__(self, main=None):
         self.mainwindow = None
         self.ismaximized = False
         self.isvisible = False
+        self.options_button = create_toolbutton(self, text=_('Options'),
+                                                icon=ima.icon('tooloptions'))
+        self.options_button.setPopupMode(QToolButton.InstantPopup)
+        self.menu = QMenu(self)
```
Comment (ccordoba12) [inline spyder/api/plugins.py]: `menu` is too generic. Please change it to `options_menu`.

**Comment by ccordoba12** [comment]
@dalthviz, I left some simple comments, then I'll merge.

**Comment by ccordoba12** [comment]
@dalthviz, please squash the last two commits into one with the message "Fix tests".

**Pull request merged**